toolkit/tools/internal/packagerepo/repocloner/rpmrepocloner/rpmrepocloner.go

```diff
@@ -615,47 +622,92 @@ func (r *RpmRepoCloner) clonePackage(baseArgs []string) (preBuilt bool, err erro
-		for _, line := range splitStdout {
-			trimmedLine := strings.TrimSpace(line)
-			// Toybox conflicts are a known issue, reset the err value if encountered
-			if strings.HasPrefix(trimmedLine, toyboxConflictsPrefix) {
-				logger.Log.Warn("Ignoring known toybox conflict")
-				err = nil
-				continue
-			}
-			// If a package was not available, update err
-			if strings.HasPrefix(trimmedLine, unresolvedOutputPrefix) && strings.HasSuffix(trimmedLine, unresolvedOutputPostfix) {
-				err = fmt.Errorf(trimmedLine)
-				break
-			}
-		}
+	return
+}
 
-		if err == nil {
-			preBuilt = r.reposArgsHaveOnlyLocalSources(reposArgs)
+func tdnfDownload(args ...string) (err error, retriable bool) {
+	const (
+		unresolvedOutputPrefix  = "No package"
+		toyboxConflictsPrefix   = "toybox conflicts"
+		unresolvedOutputPostfix = "available"
+	)
+
+	var (
+		stdout string
+		stderr string
+	)
+	stdout, stderr, err = shell.Execute("tdnf", args...)
+
+	logger.Log.Debugf("stdout: %s", stdout)
+	logger.Log.Debugf("stderr: %s", stderr)
+
+	if err != nil {
+		logger.Log.Debugf("tdnf error (will continue if the only errors are toybox conflicts):\n '%s'", stderr)
+	}
+
+	// ============== TDNF SPECIFIC IMPLEMENTATION ==============
+	//
+	// Check if TDNF could not resolve a given package. If TDNF does not find a requested package,
+	// it will not error. Instead it will print a message to stdout. Check for this message.
+	//
+	// *NOTE*: TDNF will attempt best effort. If N packages are requested, and 1 cannot be found,
+	// it will still download N-1 packages while also printing the message.
+	splitStdout := strings.Split(stdout, "\n")
+	for _, line := range splitStdout {
+		trimmedLine := strings.TrimSpace(line)
+		// Toybox conflicts are a known issue, reset the err value if encountered
+		if strings.HasPrefix(trimmedLine, toyboxConflictsPrefix) {
+			logger.Log.Warn("Ignoring known toybox conflict")
+			err = nil
+			continue
+		}
+		// If a package was not available, update err
+		if strings.HasPrefix(trimmedLine, unresolvedOutputPrefix) && strings.HasSuffix(trimmedLine, unresolvedOutputPostfix) {
+			err = fmt.Errorf(trimmedLine)
 			break
 		}
 	}
 
+	//
+	// *NOTE*: There are cases in which some of our upstream package repositories are hosted
+	// on services that are prone to intermittent errors (e.g., HTTP 502 errors). We
+	// specifically look for such known cases and apply some retry logic in hopes of getting
+	// a better result; note that we don't indiscriminately retry because there are legitimate
+	// cases in which the upstream repo doesn't contain the package and a 404 error is to be
+	// expected. This involves scraping through stderr, but it's better than not doing so.
+	//
+	if err != nil {
+		for _, line := range strings.Split(stderr, "\n") {
+			if strings.Contains(line, "Error: 502 when downloading") {
+				logger.Log.Warn("Encountered possibly intermittent HTTP 502 error.")
+				retriable = true
+			}
+		}
```

Comment:
Two things:
1. Do we only want to deal with 502s or all 5XX errors? Since we're here, it might be worth handling all server-side issues.
2. What do you think about dropping the `for` loop in favour of a regular expression?

```suggestion
		serverErrorMatch := serverErrorsRegex.FindStringSubmatch(stderr)
		if len(serverErrorMatch) > errorCodeIndex {
			logger.Log.Warn("Encountered possibly intermittent HTTP %d error.", serverErrorMatch[errorCodeIndex])
			retriable = true
		}
```

The regex and the index could be defined as global, private variables (making it global forces a one-time compilation of the regex during program's start-up):
```golang
// In global context:
var (
(...)
serverErrorsRegex = regexp.MustCompile(`(?m)Error: (5\d{2}) when downloading`)
errorCodeIndex = 1
...
```